### PR TITLE
Added a progressbar during loading + threaded loading for systems

### PR DESCRIPTION
**Pull request by fabricecaruso72**
- Load systems using threaded loading when processor threads > 2 (so it doesn't apply on pi0), making loading systems faster.
- Added a progressbar during loading.
- Windows : added "sdl event dispatch" to avoid Windows thinking the app not responding during loading.

### CODE REVIEW

**Comment by escalade** [comment]
Does not seem to respect the --no-splash commandline option or the setting bools SplashScreen and SplashScreenProgress.

**Comment by fabricecaruso** [comment]
> 
> 
> Does not seem to respect the --no-splash commandline option or the setting bools SplashScreen and SplashScreenProgress.

Well seen. I've fixed this (See the call to loadSystemConfigFile)

**Comment by shantigilbert** [comment]
This is a very welcomed update! ES loading time reduced from ~3 to ~1 minute! (I have LOTS of ROMS with video previews etc) Thanks @fabricecaruso

**Comment by 5schatten** [comment]
@fabricecaruso 
this PR needs a rebase

**Comment by shantigilbert** [comment]
This PR no longer works on current master:

```
../es-core/src/Window.cpp: In member function 'void Window::renderLoadingScreen(std::string, float, unsigned char)':
../es-core/src/Window.cpp:317:54: error: no matching function for call to 'drawRect(float&, float&, float&, float&, int)'
   Renderer::drawRect(x, y, w, h, 0x25252500 | opacity);

../es-core/src/renderers/Renderer.h:70:14: note:   candidate expects 9 arguments, 5 provided
../es-core/src/Window.cpp:318:64: error: no matching function for call to 'drawRect(float&, float&, float, float&, int)'
   Renderer::drawRect(x, y, (w*percent), h, 0x006C9E00 | opacity); // 0xFFFFFFFF

```

**Comment by 5schatten** [comment]
@tomaz82 
Can you have a look at this? Works fine here for a while for Generic & RPi2/3 builds.

**Comment by shantigilbert** [comment]
It can be fixed by changing these lines in /es-core/src/Window.cpp: 

```
Renderer::drawRect(x, y, w, h, 0x25252500 | opacity);
Renderer::drawRect(x, y, (w*percent), h, 0x006C9E00 | opacity); // 0xFFFFFFFF
```

to : 
```
Renderer::drawRect(x, y, w, h, 0x25252500, 0x25252500 | opacity);
Renderer::drawRect(x, y, (w*percent), h, 0x006C9E00, 0x006C9E00 | opacity); // 0xFFFFFFFF
```

Although I am not 100% sure this is the correct fix, but it compiled and it works for me

**Comment by tomaz82** [comment]
> It can be fixed by changing these lines in /es-core/src/Window.cpp:
> 
> ```
> Renderer::drawRect(x, y, w, h, 0x25252500 | opacity);
> Renderer::drawRect(x, y, (w*percent), h, 0x006C9E00 | opacity); // 0xFFFFFFFF
> ```
> 
> to :
> 
> ```
> Renderer::drawRect(x, y, w, h, 0x25252500, 0x25252500 | opacity);
> Renderer::drawRect(x, y, (w*percent), h, 0x006C9E00, 0x006C9E00 | opacity); // 0xFFFFFFFF
> ```
> 
> Although I am not 100% sure this is the correct fix, but it compiled and it works for me

That is not correct, the 2 last variables should be identical ( add | opacity ), it's the start/end colors when doing gradients, when no gradient is wanted both variables should be identical.

**Comment by tomaz82** [comment]
> @tomaz82
> Can you have a look at this? Works fine here for a while for Generic & RPi2/3 builds.

Take a look at what? Review this PR?

**Comment by fabricecaruso** [comment]
Ok, rebased & fixed

**Comment by 5schatten** [comment]
@tomaz82 
Review or merge would be great _if_ it's ready. Because it's a bit annoying that the backport patch breaks every time something else is merged before this PR :see_no_evil:

**Comment by shantigilbert** [comment]
@fabricecaruso is this correct?

 ```
+	if (percent >= 0)
+	{
+		float baseHeight = 0.04f;
+
+		float w = Renderer::getScreenWidth() / 2;
+		float h = Renderer::getScreenHeight() * baseHeight;
+
+		float x = Renderer::getScreenWidth() / 2 - w / 2;
+		float y = Renderer::getScreenHeight() - (Renderer::getScreenHeight() * 3 * baseHeight);
+
+		Renderer::drawRect(x, y, w, h, 0x25252500 | opacity, 0x25252500 | opacity);
+		Renderer::drawRect(x, y, (w*percent), h, 0x006C9E00 | opacity, 0x006C9E00 | opacity); // 0xFFFFFFFF
+	}
+
+	if (percent >= 0)
+	{
+		float baseHeight = 0.04f;
+
+		float w = Renderer::getScreenWidth() / 2;
+		float h = Renderer::getScreenHeight() * baseHeight;
+
+		float x = Renderer::getScreenWidth() / 2 - w / 2;
+		float y = Renderer::getScreenHeight() - (Renderer::getScreenHeight() * 3 * baseHeight);
+
+		Renderer::drawRect(x, y, w, h, 0x25252500 | opacity, 0x25252500 | opacity);
+		Renderer::drawRect(x, y, (w*percent), h, 0x006C9E00 | opacity, 0x006C9E00 | opacity); // 0xFFFFFFFF
+	}
```

seems to be doubled

**Comment by fabricecaruso** [comment]
Strange, don't know how the block was duplicated. It is fixed now

**Comment by joolswills** [comment]
@tomaz82 - you ok for this to be merged?

File: es-app/src/SystemData.cpp [hunk 1]
```diff
@@ -186,7 +272,7 @@ bool SystemData::loadConfig()
 	pugi::xml_document doc;
 	pugi::xml_parse_result res = doc.load_file(path.c_str());
 
-	if(!res)
+	if (!res)
 	{
```
Comment (joolswills) [inline es-app/src/SystemData.cpp]: I noticed this and other whitespace changes. Although the formatting of es isn't ideal, this shouldn't be done as part of this PR. Please can you include code changes only.

File: es-app/src/SystemData.h [hunk 2]
```diff
@@ -81,7 +86,7 @@ class SystemData
 	void populateFolder(FileData* folder);
 	void indexAllGameFilters(const FileData* folder);
 	void setIsGameSystemStatus();
-
+
```
Comment (joolswills) [inline es-app/src/SystemData.h]: Whitespace

File: es-app/src/SystemData.cpp [hunk 3]
```diff
@@ -196,86 +282,104 @@ bool SystemData::loadConfig()
 	//actually read the file
 	pugi::xml_node systemList = doc.child("systemList");
 
-	if(!systemList)
+	if (!systemList)
```
Comment (joolswills) [inline es-app/src/SystemData.cpp]: As above

**Comment by joolswills** [comment]
This will need a rebase now. Can't remember if this was before, or supplements the progress bar which was merged previously.

**Comment by steeeb** [comment]
This worked well (Odroid XU4) - had to also apply 0d3bbdec296e94870df8341e60af33b53b4da829 (Undo change to use std::array introduced by #646) to get it to build though

**Comment by SupervisedThinking** [comment]
@fabricecaruso could you have a look at this and rebase it agains latest stable? 🥇

**Comment by SupervisedThinking** [comment]
@fabricecaruso72 thx for rebasing but could you rebase against? https://github.com/RetroPie/EmulationStation/tree/stable otherwise patches v2.9.4 fails

File: es-core/src/Window.cpp [hunk 4]
```diff
@@ -9,6 +9,7 @@
 #include "Scripting.h"
 #include <algorithm>
 #include <iomanip>
+#include <SDL_events.h>
```
Comment (pjft) [inline es-core/src/Window.cpp:12]: This should also be around a #if defined(_WIN32) , correct?

**Comment by pjft** [comment]
I have no objection, would rather have the SDL_Event be included only when it makes sense (hence my comment).

**Comment by joolswills** [comment]
Just to note PRs are to be merged into master. Not sure why it was requested to be rebased over stable.

**Comment by SupervisedThinking** [comment]
@joolswills probably because I want to use it *now* with stable and since this PR is around since more than 1,5 years it should have been merged into stable for quite some time.

**Comment by pjft** [comment]
"having been around for more than 1,5 years" and "should have been merged into stable for quite some time" aren't necessarily correlated.

Things are merged when reviewed and tested.

Things are moved to stable when having been used by the user base in emulationstation-dev for a while.

**Comment by SupervisedThinking** [comment]
@pjft https://github.com/RetroPie/EmulationStation/pull/590#issuecomment-532039386 it's not like it's wasn't reviewed

**Comment by pjft** [comment]
@fabricecaruso72 Apologies for the delay in getting to this. I am comfortable with these changes at first glance - I just briefly tested them - but before we merge them, could you please:

- Make sure it is rebased against master (it seems to be, but the latest messages I'm not clear of the current status);
- Address the previous remarks from @joolswills 's review;
- Squash them in a single commit?

Thanks.

**Comment by SupervisedThinking** [comment]
@pjft it's squashed, rebased, tested for years... but still not yet merged. I guess you might understand now why some people start to nag about it 🙄

**Comment by SupervisedThinking** [comment]
@fabricecaruso72 could you check you PR if it finds games with capital extensions like `<extension>.BIN .self</extension>`? Looks like it's actually skipping them.

**Comment by fabricecaruso** [comment]
You're right. Fixed & squashed.

**Comment by SupervisedThinking** [comment]
@fabricecaruso 🥇 
could you create a branch in your repo which is based on stable? So you don't have rebase it again and again between stable & master & I would just grab it from there?

**Comment by joolswills** [comment]
It doesn't need to be rebased back and forth. If you want to use it against the stable branch you can rebase it yourself.

**Comment by SupervisedThinking** [comment]
@joolswills well if you read again I was asking for a branch in his repo and not a rebased PR. If you don't merge this PR, for whatever reason, it's one thing. But some people might want to use it on top of the current stable version.

**Comment by pjft** [comment]
Thanks @fabricecaruso72 for sending this over.

**Pull request merged**